nbviewer/providers/local/handlers.py

```diff
@@ -88,6 +89,42 @@ def download(self, abspath):
             except iostream.StreamClosedError:
                 return
 
+    def can_show(self, path):
+        """
+        Generally determine whether the given path is displayable.
+        This function is useful for failing fast - further checks may
+        be applied at notebook render to confirm a file may be shown.
+
+        """
+        abspath = os.path.realpath(os.path.normpath(os.path.join(
+            self.localfile_path,
+            path
+        )))
+
+        if not abspath.startswith(self.localfile_path):
+            app_log.warn("directory traversal attempt: '%s'" %
+                         abspath)
+            return False
+
+        if not os.path.exists(abspath):
+            return False
+
+        if any(part.startswith('.') or part.startswith('_')
+               for part in abspath.split(os.sep)):
+            return False
+
+        fstat = os.stat(abspath)
+
+        # Ensure the file/directory has other read access for all.
+        if not fstat.st_mode & stat.S_IROTH:
+            return False
+
+        if os.path.isdir(abspath) and not fstat.st_mode & stat.S_IXOTH:
```

Comment:
Learning moment (for me): is there a reason to switch to `fstat` and `stat` from the `os.access` calls used in the original?

Reply:
The big difference (that is quite opinionated) is that before we were checking if the **user** was able to read a file, where now we are checking that **all users** are able to read a file. ```os.access``` only allows you to test the permissions for the user.

To explain my usecase further: I'm running nbviewer under its own dedicated user, and don't want any back-doors to allow people to list/cat files that aren't visible to everybody anyway. Doing so means I no longer need to worry about somebody getting hold of things like ssh keys etc., and prevents filesystem permissions user-hopping. I would argue that this change in behaviour is a much *safer* default.

Reply:
Thanks for explaining your use case. We, too, run a local nbviewer in a docker container with select volumes mounted from the host to control what gets shared and what doesn't. I agree that switching to other readable/listable checks only increases security.